### PR TITLE
feat(flags): Add --absolute flag

### LINKED ISSUE #726: feat:  display  file/folder names with full path
It would be nice to have a flag to toggle the display of full paths with file and folder names.

Something similar to `find $PWD | xargs ls -ld`

You don't need this often, but sometimes such behaviour is desired if you plan to copy/paste file names to another terminal window, for example.

### PR DESCRIPTION
**Pull request by taminob**
##### Description

This feature allows to display the full path of files and directories when passing the --absolute flag.
I additionally allowed users to only print the full path for files or directories, but it can also be easily converted back to a boolean flag.

For now, the flag behaves the same for all modes which makes most sense in my opinion.
However, the combination with `--tree` can lead to quite verbose output.

##### How Has This Been Tested?

I tested the following cases manually for regular files, directories and symlinks:

- `eza` without arguments and with `--absolute={,on,follow,off}`
- above in combination with `--long`, `--tree`, `--recurse`, `--classify`, `--hyperlink`
- `eza` zsh shell completions

Additionally, I added some basic integration tests to `tests/cmd`.
Since there isn't any complex logic involved in the added changes, I'm not sure if unit tests make sense. Only thing worth it would probably be the `absolute_path`, but I would appreciate some hints on how to mock the filesystem in Rust tests.

Resolves #726

### CODE REVIEW

**Comment by taminob** [comment]
Feel free to nitpick as much as you want, my last time working with Rust has been some time ago, so there are probably better ways to express the same in a more concise way. Also not entirely happy with the naming of the `display_name` function, suggestions are welcome.

File: src/options/file_name.rs [hunk 1]
```diff
@@ -108,3 +113,23 @@ impl EmbedHyperlinks {
         }
     }
 }
+
+impl Absolute {
+    fn deduce(matches: &MatchedFlags<'_>) -> Result<Self, OptionsError> {
+        let mode_opt = matches.get(&flags::ABSOLUTE)?;
+        if !matches.has(&flags::ABSOLUTE)? && mode_opt.is_none() {
+            return Ok(Self::Nothing);
```
Comment (MartinFillon) [inline src/options/file_name.rs:121]: Nitpick but unuseful condition there. (I know this is how its done on the other parts of the code but I am trying, slowly, to remove that type of code)

**Review by MartinFillon — DISMISSED** [review]
If you could also modify the powertests.yaml file it would be great

**Comment by taminob** [comment]
> If you could also modify the powertests.yaml file it would be great

I added it to that file, but how do I generate the files in `tests/ptests`? Didn't find any documentation regarding that.
In case it requires a Nix setup, do you have a development docker image based on Nix with the dependencies installed?

**Comment by MartinFillon** [comment]
The goal of nix is to not use docker and if you wanna regenerate please see powertest repo or I will do it once this is merged

**Comment by taminob** [comment]
> The goal of nix is to not use docker and if you wanna regenerate please see powertest repo or I will do it once this is merged

Got it, sorry - I confused NixOS and Nix. I didn't know it's possible to run it outside of that distro as well. Guess I have some reading to do.

I added the generated files, although I appended a newline at the end of them. For me, running powertest removed all the newlines for all the files in `tests/ptests` - not sure if this is a bug or there was some mistake in my setup.

**Comment by MartinFillon** [comment]
not a problem but lets keep it like that tho, last thing todo (I am starting to see that we lack doc on that I am going to make it I think) is running `just idump`

**Comment by taminob** [comment]
> not a problem but lets keep it like that tho, last thing todo (I am starting to see that we lack doc on that I am going to make it I think) is running `just idump`

Mentioning the `idump` target in `CONTRIBUTING.md` wouldn't hurt, I guess. But no need to duplicate the nix documentation - it is quite extensive, I'm just unfamiliar with it since I always thought it's bound to NixOS.
Linking the `powertest` repo could be useful for others, too, since it's not the first thing you'll find when searching for it.

Otherwise, `just idump` and `cargo test` contradict each other with their expected output - one wants `[CWD]` as root and the other one `/build/source/`. I kept `[CWD]` for now to keep the CI green.

**Comment by taminob** [comment]
Just some other thing I thought about - maybe another value for `--absolute` would make sense: `--absolute=no-follow` since the currently used function for the absolute path will resolve all symlinks on the path which is not necessarily what the user wants.
However, I think we could also add this in a follow-up PR (except you think it would be better if this is the default behavior?).

Since that kind of conflicts with the `WHAT` for `--absolute` (since it is more a `HOW`), do you want me to remove `files`/`dirs` as possible values? To be honest, I don't really see a use-case for this anymore.

**Comment by MartinFillon** [comment]
> --absolute=no-follow since the currently used function for the absolute path will resolve all symlinks on the path which is not necessarily what the user wants

What about inverting that to follow and doing it only if foollowing

**Comment by taminob** [comment]
> > --absolute=no-follow since the currently used function for the absolute path will resolve all symlinks on the path which is not necessarily what the user wants
> 
> What about inverting that to follow and doing it only if following ?

I modified the behavior to not follow symlinks in the absolute path by default and added `--absolute=follow` to follow symlinks. Do you think this is sufficiently self-explanatory or should I rename it to `--absolute=follow-symlinks`?

I also noticed that the behavior for broken vs valid symlinks might be a bit confusing because for valid symlinks the target will be displayed with `--absolute=follow` (e.g. `/path/to/symlink -> a` will be displayed as `/path/to/a` if valid but `/path/to/symlink` if broken). But not resolving the final component if possible would be kind of inconsistent. Any other thoughts?

Otherwise, the current `tests/itest` directory does not have any symlink directory (and thus the test output for `--absolute` and `--absolute=follow` is identical).
I'd propose to add that in another PR since that would touch a lot of additional files (basically all test outputs).

**Review by MartinFillon — APPROVED** [review]
Good for me, just add the entrys to the test dir ot be sure

**Review by MartinFillon — DISMISSED** [review]

**Review by cafkafk — DISMISSED** [review]
Thanks for adding this feature, very useful. Integration tests pass and nothing seems to have regressed, and testing the actual behavior of the code produces the expected results. 

The code also looks sane, and the crate introduced doesn't seem too obscure to inlucde.

Overall: LGTM, thanks for the contribution :D

**Review by cafkafk — APPROVED** [review]

**Comment by j-xella** [comment]
It would also be great to mention this option in `man/eza.1.md` file ...

**Comment by cafkafk** [comment]
> It would also be great to mention this option in `man/eza.1.md` file ...

you should feel free to submit a PR to that effect, or opening an issue to track this :)

**Pull request merged**